File: src/parser/parsers/variable.js
"use strict";

const node = require("../helpers/nodes.js");
const error = require("../helpers/error.js");
const add = require("../helpers/tree-add.js");
const { nk } = require("../helpers/enums.js");
const rollback = require("../helpers/rollback.js");
const validate = require("../helpers/validate.js");
const {
	cin,
	cnotin,
	C_NL,
	C_SPACES,
	C_LETTERS,
	C_QUOTES,
	C_VAR_IDENT,
	C_VAR_VALUE
} = require("../helpers/charsets.js");

/**
 * ----------------------------------------------------------- Parsing Breakdown
 * $variable = "value"
 *          | |       ^-EOL-Whitespace-Boundary 3.
 *          ^-^-Whitespace-Boundary 1/2.
 * ^-Sigil.
 *  ^-Name.
 *           ^-Assignment.
 *             ^-Value.
 * -----------------------------------------------------------------------------
 *
 * @param  {object} S - State object.
 * @return {undefined} - Nothing is returned.
 */
module.exports = (S) => {
	let l = S.l;
	let qchar;
	let state = "sigil";
	let N = node(nk.Variable, S);

	let char,
		pchar = "";
	for (; S.i < l; S.i++, S.column++) {
		pchar = char;
		char = S.text.charAt(S.i);

		if (cin(C_NL, char)) {
			N.end = rollback(S) && S.i;
			break; // Stop at nl char.
		}

		if (char === "#" && pchar !== "\\" && state !== "value") {
			rollback(S);
			N.end = S.i;
			break;
		}

		switch (state) {
			case "sigil":
				N.sigil.start = N.sigil.end = S.i;
				state = "name";

				break;

			case "name":
				if (!N.name.value) {
					if (cnotin(C_LETTERS, char)) error(S);

					N.name.start = N.name.end = S.i;
					N.name.value = char;
				} else {
					if (cin(C_VAR_IDENT, char)) {
						N.name.end = S.i;
						N.name.value += char;
					} else if (cin(C_SPACES, char)) {
						state = "name-wsb";
						continue;
					} else if (char === "=") {
						state = "assignment";
						rollback(S);
					} else error(S);
				}

				break;

			case "name-wsb":
				if (cnotin(C_SPACES, char)) {
					if (char === "=") {
						state = "assignment";
						rollback(S);
					} else error(S);
				}

				break;

			case "assignment":
				N.assignment.start = N.assignment.end = S.i;
				N.assignment.value = char;
				state = "value-wsb";

				break;

			case "value-wsb":
				if (cnotin(C_SPACES, char)) {
					state = "value";
					rollback(S);
				}

				break;

			case "value":
				if (!N.value.value) {
					if (cnotin(C_VAR_VALUE, char)) error(S);

					if (cin(C_QUOTES, char)) qchar = char;
					N.value.start = N.value.end = S.i;
					N.value.value = char;
				} else {
					if (qchar) {
						if (char === qchar && pchar !== "\\") state = "eol-wsb";
						N.value.end = S.i;
						N.value.value += char;
					} else {
						if (cin(C_SPACES, char) && pchar !== "\\") {
							state = "eol-wsb";
							rollback(S);
						} else {
							N.value.end = S.i;
							N.value.value += char;
						}
					}
				}

				break;

			case "eol-wsb":
				if (cnotin(C_SPACES, char)) error(S);

				break;
		}
	}

	validate(S, N);
	add(S, N);

	let value = N.value.value || "";
	value = value.substring(1, value.length - 1);
	// Unquote value. [https://stackoverflow.com/a/21873245]
	// Unquote value. [https://stackoverflow.com/a/19156197]
	// value = value.replace(/^(["'])(.+(?=\1$))\1$/, "$2");

	S.tables.variables[N.name.value] = value; // Store var/val.
};

File: src/parser/nim/parser/parsers/variable.nim
import std/tables

import ../helpers/[tree_add, types, charsets]
import ../helpers/[error, validate, forward, rollback]

# ------------------------------------------------------------ Parsing Breakdown
# $variable = "value"
#          | |       ^-EOL-Whitespace-Boundary 3.
#          ^-^-Whitespace-Boundary 1/2.
# ^-Sigil.
#  ^-Name.
#           ^-Assignment.
#             ^-Value.
# ------------------------------------------------------------------------------
#
# @param  {object} S - State object.
# @return - Nothing is returned.
proc p_variable*(S: State) =
    let text = S.text
    var qchar: char
    var state = "sigil"
    var N = node(nkVariable, S)

    let l = S.l; var c, p: char
    while S.i < l:
        p = c
        c = text[S.i]

        if c in C_NL:
            rollback(S)
            N.stop = S.i
            break # Stop at nl char.

        if c == '#' and p != '\\' and state != "value":
            rollback(S)
            N.stop = S.i
            break

        case (state):
            of "sigil":
                N.sigil.start = S.i
                N.sigil.stop = S.i
                state = "name"

            of "name":
                if N.name.value == "":
                    if c notin C_LETTERS: error(S, currentSourcePath)

                    N.name.start = S.i
                    N.name.stop = S.i
                    N.name.value = $c
                else:
                    if c in C_VAR_IDENT:
                        N.name.stop = S.i
                        N.name.value &= $c
                    elif c in C_SPACES:
                        state = "name-wsb"
                        forward(S)
                        continue
                    elif c == '=':
                        state = "assignment"
                        rollback(S)
                    else: error(S, currentSourcePath)

            of "name-wsb":
                if c notin C_SPACES:
                    if c == '=':
                        state = "assignment"
                        rollback(S)
                    else: error(S, currentSourcePath)

            of "assignment":
                N.assignment.start = S.i
                N.assignment.stop = S.i
                N.assignment.value = $c
                state = "value-wsb"

            of "value-wsb":
                if c notin C_SPACES:
                    state = "value"
                    rollback(S)

            of "value":
                if N.value.value == "":
                    if c notin C_VAR_VALUE: error(S, currentSourcePath)

                    if c in C_QUOTES: qchar = c
                    N.value.start = S.i
                    N.value.stop = S.i
                    N.value.value = $c
                else:
                    if qchar != '\0':
                        if c == qchar and p != '\\': state = "eol-wsb"
                        N.value.stop = S.i
                        N.value.value &= $c
                    else:
                        if c in C_SPACES and p != '\\':
                            state = "eol-wsb"
                            rollback(S)
                        else:
                            N.value.stop = S.i
                            N.value.value &= $c

            of "eol-wsb":
                if c notin C_SPACES: error(S, currentSourcePath)

            else: discard

        forward(S)

    discard validate(S, N)
    add(S, N)

    var value = if N.value.value != "": N.value.value else: ""
    if value.len > 0 and value[0] in C_QUOTES: value = value[1 .. ^2]

    S.tables.variables[N.name.value] = value # Store var/val.
